packages/code-signing-certificates/src/index.ts

```diff
@@ -0,0 +1,144 @@
+    return hexString;
+  }
+  mostSiginficativeHexAsInt -= 8;
+  return mostSiginficativeHexAsInt.toString() + hexString.substring(1);
+}
+
+export type CertificateAttrs = PKI.CertificateField[];
+
+export function generateKeyPair(): PKI.rsa.KeyPair {
+  return PKI.rsa.generateKeyPair();
+}
+
+export function convertKeyPairToPEM(
+  keyPair: PKI.rsa.KeyPair
+): { privateKeyPEM: string; publicKeyPEM: string } {
+  return {
+    privateKeyPEM: PKI.privateKeyToPem(keyPair.privateKey),
+    publicKeyPEM: PKI.publicKeyToPem(keyPair.publicKey),
+  };
+}
+
+export function convertCertificateToPEM(cert: PKI.Certificate): string {
+  return PKI.certificateToPem(cert);
+}
+
+type GenerateParameters = {
+  /**
+   * Public/private key pair generated via {@link generateKeyPair}.
+   */
+  keyPair: PKI.rsa.KeyPair;
+
+  /**
+   * Certificate validity range start.
+   */
+  validityNotBefore: Date;
+
+  /**
+   * Certificate validity range end.
+   */
+  validityNotAfter: Date;
+
+  /**
+   * CN issuer and subject Distinguished Name (DN).
+   * Used for both issuer and subject in the case of self-signed certificates.
+   */
+  commonName: string;
+};
+
+/**
+ * Generate a self-signed code-signing certificate for use with expo-updates.
+ * Note that while certificate chains may be supported at some point in expo-updates, for now
+ * only self-signed certificates are supported.
+ *
+ * @returns PKI.Certificate valid for expo-updates code signing
+ */
+export function generateSelfSignedCodeSigningCertificate({
+  keyPair: { publicKey, privateKey },
+  validityNotBefore,
+  validityNotAfter,
+  commonName,
+}: GenerateParameters): PKI.Certificate {
+  const cert = PKI.createCertificate();
+  cert.publicKey = publicKey;
+  cert.serialNumber = toPositiveHex(util.bytesToHex(random.getBytesSync(9)));
+
+  assert(
+    validityNotAfter >= validityNotBefore,
+    'validityNotAfter must be later than validityNotBefore'
+  );
+  cert.validity.notBefore = validityNotBefore;
+  cert.validity.notAfter = validityNotAfter;
+
+  const attrs = [
+    {
+      name: 'commonName',
+      value: commonName,
+    },
+  ];
+  cert.setSubject(attrs);
+  cert.setIssuer(attrs);
+
+  cert.setExtensions([
+    {
+      name: 'keyUsage',
+      critical: true,
+      keyCertSign: false,
+      digitalSignature: true,
+      nonRepudiation: false,
+      keyEncipherment: false,
+      dataEncipherment: false,
+    },
+    {
+      name: 'extKeyUsage',
+      critical: true,
+      serverAuth: false,
+      clientAuth: false,
+      codeSigning: true,
+      emailProtection: false,
+      timeStamping: false,
+    },
+  ]);
+
+  cert.sign(privateKey, md.sha256.create());
+  return cert;
+}
+
+/**
+ *
+ * @param certificate
+ */
+export function validateSelfSignedCertificate(certificate: PKI.Certificate) {
+  if (certificate.issuer.hash !== certificate.subject.hash) {
+    throw new Error(
+      'Certificate issuer has does not match subject hash, certificate may not be self-signed.'
```

Comment:
```suggestion
      'Certificate issuer hash does not match subject hash, certificate may not be self-signed.'
```
Also a bit confused, the function name says "validate self signed" the error message seems to demand that it is not?

Reply:
Haha, english is bad and should feel bad. I meant "may" more as the "expressing possibility" version of the word than the "expressing permission" version. Will update to be clearer.